R/standardize_treatment.R

```diff
@@ -0,0 +1,445 @@
+#' @param target Vector of population means to re-weight to
+#' @param Z Vector of group indicators
+#' @param S_factor Vector of site indicators
+#' @param pro_trt_split List of J vectors of treatment propensity multipliers
+#' @param pro_ctr_split List of J vectors of control propensity multipliers
+#' @param lowlim Lower limit on weights
+#' @param uplim Upper limit on weights
+#' @param exact_global Boolean indicating whether to enforce global constraint
+#' @param verbose Boolean indicating whether to print progress
+#'
+#' @return A, l, and u
+create_constraints_treatment <- function(X0s, Xtaus, target, Z, S_factor,
+                               pro_trt_split, pro_ctr_split,
+                               lowlim, uplim, exact_global, verbose) {
+
+  J <- length(X0s)
+  nj <- as.numeric(sapply(X0s, nrow))
+  d0 <- ncol(X0s[[1]])
+  dtau <- ncol(Xtaus[[1]])
+  cumsum_nj <- cumsum(c(1, nj))
+  n <- sum(nj)
+  X0st <- lapply(X0s, t)
+  Xtaust <- lapply(Xtaus, t)
+
+  # compute number of treated and control units within each site
+  n1j <- sapply(1:J, FUN = function(j) sum(Z[S_factor == j]))
+  n0j <- sapply(1:J, FUN = function(j) sum(1 - (Z[S_factor == j])))
+
+  # dimension of auxiliary weights
+  aux_dim <- (J * d0) + (J * dtau)
+
+  if(verbose) message("\tx Sum to one constraint")
+  # sum-to-one constraint for each group
+  trt_mat <- Matrix::t(Matrix::bdiag(split(Z, S_factor)))
+  ctr_mat <- Matrix::t(Matrix::bdiag(split(1 - Z, S_factor)))
+  A1 <- Matrix::rbind2(trt_mat, ctr_mat)
+
+  rm(trt_mat)
+  rm(ctr_mat)
+
+  A1 <- Matrix::cbind2(A1, Matrix::Matrix(0, nrow=nrow(A1), ncol = aux_dim))
+  l1 <- c(n1j, n0j)
+  u1 <- c(n1j, n0j)
+
+  if(verbose) message("\tx Upper and lower bounds")
+  # upper and lower bounds
+  A2 <- Matrix::Diagonal(n)
+  A2 <- Matrix::cbind2(A2, Matrix::Matrix(0, nrow = nrow(A2), ncol = aux_dim))
+  l2 <- rep(lowlim, n)
+  u2 <- rep(uplim, n)
+
+  if(exact_global) {
+    if(verbose) message("\tx Mantain overall population mean")
+    # require that the weighted average of the treated units (over all sites)
+    # equals the weighted average of the control units (over all sites)
+    response_mat <- t(
+                      t(
+                        do.call(cbind, X0st)) * ((unlist(pro_trt_split) * rep(n1j / sum(n1j), nj)) -
+                                                 (unlist(pro_ctr_split) * rep(n0j / sum(n0j), nj))))
+    treat_mat <- t(t(do.call(cbind, Xtaust)) * rep(n1j, nj) * unlist(pro_trt_split))
+
+    A3 <- Matrix::rbind2(response_mat, treat_mat)
+
+    rm(response_mat)
+    rm(treat_mat)
+
+    A3 <- Matrix::cbind2(A3, Matrix::Matrix(0, nrow = nrow(A3), ncol = aux_dim))
+
+    l3 <- c(rep(0, d0), sum(n1j) * target)
+    u3 <- c(rep(0, d0), sum(n1j) * target)
+  } else {
+    if(verbose) message("\t(SKIPPING) Mantain overall population mean")
+    # skip this constraint and just make empty
+    A3 <- matrix(, nrow = 0, ncol = ncol(A2))
+    l3 <- numeric(0)
+    u3 <- numeric(0)
+  }
```

Comment:
I would just drop the exact global constraint entirely, it's not needed here.

Reply:
Done.